New segment filter purchased with attribute [MAILPOET-5467]

johnolek opened the pull request:
## Description

This PR adds a new dynamic segment filter for "purchased with attribute".

Attributes in WooCommerce can be either global, which are saved as taxonomy/terms, or local, which are specific to a product. This filter accounts for both types, but with the limitation that it will only look at local attributes that were used to create product variations.

When local attributes are used for product variations, a new row is created in the `postmeta` table for each attribute/value pair, which makes it possible to write efficient queries to target them. If a simple product has multiple values for a local attribute, it's also saved in the `postmeta` table, but as a serialized string like this:

```
a:1:{s:15:"local-attribute";a:6:{s:4:"name";s:15:"Local Attribute";s:5:"value";s:31:"One | Two | Three | Four | Five";s:8:"position";i:0;s:10:"is_visible";i:1;s:12:"is_variation";i:0;s:11:"is_taxonomy";i:0;}}
```

Since the `postmeta` table is likely to be massive on any larger site and the `meta_value` column is not indexed, I don't see a way to query for this data in an efficient way.

I don't see this as a major issue because if a product has some unique local attributes, a user could simply use the "purchased product" filter.

## Code review notes

_N/A_

## QA notes

Please make sure to test both global (taxonomy based) and local attributes. 

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-5467](https://mailpoet.atlassian.net/browse/MAILPOET-5467)

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5467]: https://mailpoet.atlassian.net/browse/MAILPOET-5467?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

rodrigoprimo committed:
Add `purchased with attribute` segment UI code

[MAILPOET-5467]

Commit:
Make attribute term ids a multiselect

MAILPOET-5467

Commit:
Use taxonomy slug instead of id

The lookup table, wc_product_attributes_lookup, that we'll use for the
queries, includes the slug but not the taxonomy id.

MAILPOET-5467

Commit:
Add backend logic for purchase by attribute filter

MAILPOET-5467

Commit:
Add tracking for purchased with attribute filter

MAILPOET-5467

Commit:
Ensure selected terms appear in dropdown on load

MAILPOET-5467

Commit:
Only load WC data when WC is active

MAILPOET-5467

Commit:
Add filter data validation

MAILPOET-5467

Commit:
Add explanation of the getLookupData method

MAILPOET-5467

Commit:
Add support for local variation attributes

In WooCommerce it's possible to create attributes that are local to a
specific product. When these attributes are used to generate variations,
 they are stored in the postmeta table.

MAILPOET-5467

costasovo commented on mailpoet/assets/js/src/segments/dynamic/dynamic-segments-filters/fields/woocommerce/purchased-with-attribute.tsx:
```diff
@@ -0,0 +1,137 @@
+      value: term.term_id,
+      label: term.name,
+    }));
+  }, [segment.attribute_id, productAttributes]);
```
Maybe I'm missing something, but to me, it seems that here we want to "cache" the mapped value of product attributes between renders. Am I correct?
I would say that more natural and probably also more efficient way of doing this could using `useMemo` (https://react.dev/reference/react/useMemo)

Could look like this (not tested):
```js

const productAttributeTermsOptions = useMemo(() => {
  if (segment.attribute_id === undefined) {
     return null;
  }
  return productAttributes[
    segment.attribute_id
  ].terms.map((term) => ({
    value: term.term_id,
    label: term.name,
  })
}, [segment.attribute_id, productAttributes]);

johnolek replied:
Thanks for this comment. I wasn't sure what the intention was with the original `useRef` myself. I've refactored this to use `useMemo`.

costasovo commented on mailpoet/assets/js/src/segments/dynamic/dynamic-segments-filters/fields/woocommerce/purchased-with-attribute.tsx:
```diff
@@ -0,0 +1,137 @@
+    }));
+  }, [segment.attribute_id, productAttributes]);
+
+  useEffect(() => {
+    if (
+      segment.operator !== AnyValueTypes.ALL &&
+      segment.operator !== AnyValueTypes.ANY &&
+      segment.operator !== AnyValueTypes.NONE
+    ) {
+      void updateSegmentFilter({ operator: AnyValueTypes.ANY }, filterIndex);
+    }
+  }, [updateSegmentFilter, segment, filterIndex]);
```
I've recently read a nice post about cases where `useEffect` might not be the best solution. I think this might be one of those cases ([Adjusting some state when props changes](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes)). Using the advice from the post could save us a render. But it is not a big deal.

johnolek replied:
Thanks for the link! I'm going to keep this as is for now since it's such a common pattern in this part of the codebase.

costasovo commented on mailpoet/lib/Segments/DynamicSegments/FilterDataMapper.php:
```diff
@@ -509,6 +510,10 @@ private function createWooCommerce(array $data): DynamicSegmentFilterData {
+    } elseif ($data['action'] === WooCommercePurchasedWithAttribute::ACTION) {
+      $filterData['operator'] = $data['operator'];
+      $filterData['attribute_taxonomy_slug'] = $data['attribute_taxonomy_slug'];
+      $filterData['attribute_term_ids'] = $data['attribute_term_ids'];
```
Maybe we could add a couple of validations like we have for other filter types ([e.g.](https://github.com/mailpoet/mailpoet/pull/5437/commits/90e35c24325934cf629a34b46df5c8b55211edc2#diff-44dc56dc4354bb109ce8925b08581157b0d1a381f2e399d20204d4ea4fe40455R494-R503)).
This could prevent saving invalid data from the clients. What do you think?

Please note that there is also [a test](https://github.com/mailpoet/mailpoet/blob/trunk/mailpoet/tests/integration/Segments/DynamicSegments/FilterDataMapperTest.php) for this class so it would be nice to add cases for this filter type.

johnolek replied:
Good point about the missing tests for the class.  I added some here: https://github.com/mailpoet/mailpoet/tree/e030c017c6cd57be5a502e7c0d8dcfe902239f29

I did have validations before [here](https://github.com/mailpoet/mailpoet/blob/e030c017c6cd57be5a502e7c0d8dcfe902239f29/mailpoet/lib/Segments/DynamicSegments/FilterDataMapper.php#L518-L518). I've found it nicer to include the validation logic in the filter class itself instead of in the data mapper. I think it's [more straightforward to test](https://github.com/mailpoet/mailpoet/blob/e030c017c6cd57be5a502e7c0d8dcfe902239f29/mailpoet/tests/integration/Segments/DynamicSegments/Filters/WooCommercePurchasedWithAttributeTest.php#L236-L328) and it's useful to run the validations in the `apply` as well to be sure that we have all the necessary data before building the query. Let me know if you think there's a better way to handle this though.

costasovo replied:
Having it together with a filter definition makes sense, and I agree it is a better place than in the mapper. My concern is that we would skip validation on save and allow saving invalid combinations (maybe due to a change after the JS dependencies update, we may start sending wrong data). But I might be missing something. Are the data validated somehow before saving?

johnolek replied:
@costasovo The validation is running in the mapper as well, currently [here](https://github.com/mailpoet/mailpoet/blob/a870a2f3edec650b70fe3406e5bdcce6bf380c03/mailpoet/lib/Segments/DynamicSegments/FilterDataMapper.php#L518-L518). I don't know of anywhere else we would need to run the validation before saving.

costasovo replied:
Ahh! Perfect! So I missed that. Thank you!

costasovo commented on mailpoet/assets/js/src/segments/dynamic/dynamic-segments-filters/fields/woocommerce/purchased-with-attribute.tsx:
```diff
@@ -90,23 +137,52 @@ export function PurchasedWithAttributeFields({
-        }, productAttributesOptions)}
+        options={combinedOptions}
+        value={
+          segment.attribute_type === 'local'
+            ? filter((localAttributeOption) => {
+                if (!segment.attribute_local_name) {
+                  return undefined;
+                }
+                return (
+                  `${segment.attribute_local_name}@local` ===
+                  localAttributeOption.value
+                );
+              }, localAttributeOptions)
+            : filter((productAttributeOption) => {
+                if (segment.attribute_taxonomy_slug === undefined) {
+                  return undefined;
+                }
+                return (
+                  segment.attribute_taxonomy_slug ===
+                  productAttributeOption.value
+                );
+              }, productAttributesOptions)
+        }
```
Maybe it could be slightly more readable to store the value into a variable before the JSX code. I think it would be easier to read (possibly also cache with useMemo) out of the JSX code, and also, the JSX would probably be more readable.

johnolek replied:
I ended up extracting most things and then memoized a lot of stuff to make sure we're avoiding re-renders. Let me know what you think.

costasovo commented on mailpoet/assets/js/src/segments/dynamic/dynamic-segments-filters/fields/woocommerce/purchased-with-attribute.tsx:
```diff
@@ -90,23 +137,52 @@ export function PurchasedWithAttributeFields({
-            },
-            filterIndex,
-          );
+          if (localAttributeValues.includes(option.value)) {
+            void updateSegmentFilter(
+              {
+                attribute_type: 'local',
+                attribute_local_name: option.value.replace(/@local$/, ''),
+                attribute_local_values: [],
+                attribute_taxonomy_slug: null,
+                attribute_term_ids: null,
+              },
+              filterIndex,
+            );
+          } else {
+            void updateSegmentFilter(
+              {
+                attribute_type: 'taxonomy',
+                attribute_local_name: null,
+                attribute_local_values: null,
+                attribute_taxonomy_slug: option.value,
+                attribute_term_ids: [],
+              },
+              filterIndex,
+            );
+          }
```
Similarly to the previous comment we could move this into a function out of the JSX code.

costasovo reviewed (commented):
@johnolek Great job! I added a few comments. All just minor details.

As this code contains a lot of TS code please rebase it on the current trunk. We merged Gutenberg Update PR and since segments use wordpress/components there might be some TS errors after the update.

Commit:
Add filter data mapper tests

MAILPOET-5467

Commit:
Refactor attribute value/onChange

This extracts some logic out of the JSX for readability and memoizes
more of the data to avoid unnecessary re-renders

MAILPOET-5467

Commit:
Refactor attribute option value generation

MAILPOET-5467

Commit:
Don't include attributes with no known values

MAILPOET-5467

Commit:
Extract initial attribute values calculation

MAILPOET-5467

Commit:
Extract attribute values onChange handler

MAILPOET-5467

Commit:
Memoize combined options to avoid re-renders

MAILPOET-5467

johnolek commented:
Thanks very much for the quick review @costasovo and the great suggestions as always! I appreciate your eye for code quality. I think I've addressed all your points so please feel free to take another look whenever you have time.

costasovo reviewed (commented):
@johnolek Thank you for addressing all the suggestions! I find it easier to read after the changes. Do you agree?

I tested it locally and it works as expected!

I noticed an issue worth fixing during segment creation.
1) Created a new segment
2) Selected "purchased with attribute" (nothing else)
3) Got an error message (see screenshot)

![Screenshot 2024-02-20 at 11 07 16](https://github.com/mailpoet/mailpoet/assets/1082140/aaadecf7-f391-422d-a948-3d4543d9945a)

It triggered the request for size calculation when the segment was not fully configured. The error in the XHR request was: `{"errors":[{"error":"bad_request","message":"The segment type is unknown."}]}`

Commit:
Require an attribute_type in validity check

MAILPOET-5467

johnolek commented:
Thanks for doing some local testing and for catching that issue, @costasovo! I pushed a fix in a870a2f3edec650b70fe3406e5bdcce6bf380c03. I think that should be everything.

costasovo reviewed (dismissed):
@johnolek Great job! 👍

veljkho commented:
Hi @johnolek 
There is conflicting file and there is no artifact build to download and test this PR. Can you please take a look. Thanks!

johnolek commented:
@veljkho I rebased and I think this should be good to go now

Aschepikov reviewed (approved):
GJ!
